enhancements/network/service-egress-traffic-steering.md

```diff
@@ -0,0 +1,284 @@
+  annotations:
+    metallb.universe.tf/address-pool: example-pool
+    k8s.ovn.org/egress-service: '{"network-hint":172.19.0.0/24}'
+spec:
+  selector:
+    app: example
+  ports:
+    - name: http
+      protocol: TCP
+      port: 8080
+      targetPort: 8080
+  type: LoadBalancer
+```
+
+3. Advertise the service from the node in charge of the service's traffic. So far the service is "broken" - it is not reachable from outside the cluster and if the pods try to send traffic outside it would probably not come back as it is SNATed to an IP which is unknown.
+We create the advertisements targetting only the node that is in charge of the service's traffic using the `nodeSelectors` field, relying on ovn-k to label the node properly.
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: example-l2-adv
+  namespace: metallb-system
+spec:
+  ipAddressPools:
+  - example-pool
+  nodeSelectors:
+  - matchLabels:
+      k8s.ovn.org/example-service-some-namespace-host: ""
+---
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisement
+metadata:
+  name: example-l2-adv
+  namespace: metallb-system
+spec:
+  ipAddressPools:
+  - example-pool
+  nodeSelectors:
+  - matchLabels:
+      k8s.ovn.org/example-service-some-namespace-host: ""
+```
+While possible to create more advertisements resources for the `example-pool`, it is the user's responsibility to make sure that the pool is advertised only by advertisements targetting the node holding the `k8s.ovn.org/<svc-name>-<svc namespace>-host: ""` label - otherwise the traffic of the service will be broken.
+### User Stories
+
+#### Story 1
+
+As a user of OpenShift, I should be able to have functioning LoadBalancer services whose backing pods exit the cluster with the service's IP from an interface that belongs to a network I specify.
+
+### API Extensions
+A new annotation `k8s.ovn.org/egress-service` is supported for LoadBalancer services.
+
+### Test Plan
+
+- Unit tests coverage.
+- E2E coverage by creating a LoadBalancer service with the proper annotations and validating that:
+  - pod to pod traffic works.
+  - external client to service works.
+  - pod to external client works and is SNATed properly.
+
+
+### Risks and Mitigations
+
+- The solution might be a bit fragile as it relies on the user to configure the external advertisement of the service manually, with certain limitations.
+
+- Using a single node to handle all of the traffic of a given service might be a bottleneck, and we will also need to try electing nodes in a way that spreads handling these kind of services between them. Failover must also be taken care of in case a node handling a service falls.
+
+- As generally a Service's purpose is to serve ingress traffic, we might be missing some corner cases when using it also to shape egress traffic behavior.
+
+- If a pod is an endpoint of multiple LoadBalancer services that request this functionality the behavior of the SNATs is undefined.
```

Comment:
If we were implementing this upstream, I might suggest that the top-level object that the user creates should be something like a `Deployment`, and then the controllers would create a corresponding `Service` and necessary MetalLB objects based on that. Then everything would be more cleanly linked together, and it wouldn't be possible to end up in this case.

Reply:
not sure I understand how that solves this case, couldn't a user still misuse labels and end up in this situation?
from the Deployment docs:
```
Do not overlap labels or selectors with other controllers (including other Deployments and StatefulSets). Kubernetes doesn't stop you from overlapping, and if multiple controllers have overlapping selectors those controllers might conflict and behave unexpectedly.
```
+ what I mentioned in the other comment

Reply:
In the implementation we should definitely error in the logs or post an event when we see this happen, as well as indicate which service IP the endpoint is currently SNAT'ing to.